pkg/zcapld/middleware.go

```diff
@@ -42,6 +49,79 @@ type InvocationExpectations struct {
 	Action         string
 }
 
+// AriesDIDKeySecrets is a secrets storage that can return did:key httpsignatures.Secrets.
+// Based on workaround suggested by library authors here: https://github.com/igor-pavlenko/httpsignatures-go/issues/5.
+type AriesDIDKeySecrets struct {
+}
+
+// Get returns a did:key secret.
+func (a *AriesDIDKeySecrets) Get(keyID string) (httpsig.Secret, error) {
+	return httpsig.Secret{
+		KeyID:     keyID,
+		Algorithm: ariesDIDKeyCustomHTTPSigAlgorithm,
+	}, nil
+}
+
+// AriesDIDKeySignatureHashAlgorithm is a custom httpsignatures.SignatureHashAlgorithm composed of
+// the aries framework's KMS and Crypto apis, and designed to work with did:key.
+// Based on workaround suggested by library authors here: https://github.com/igor-pavlenko/httpsignatures-go/issues/5.
+type AriesDIDKeySignatureHashAlgorithm struct {
+	Crypto crypto.Crypto
+	KMS    kms.KeyManager
+}
+
+// Algorithm returns this algorithm's name.
+func (a *AriesDIDKeySignatureHashAlgorithm) Algorithm() string {
+	return ariesDIDKeyCustomHTTPSigAlgorithm
+}
+
+// Create signs data with the secret.
+func (a *AriesDIDKeySignatureHashAlgorithm) Create(secret httpsig.Secret, data []byte) ([]byte, error) {
+	key, err := (&DIDKeyResolver{}).Resolve(secret.KeyID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve did:key URL %s: %w", secret.KeyID, err)
+	}
+
+	// TODO we are assuming curve: https://github.com/trustbloc/edge-core/issues/108.
+	// TODO we shouldn't be using a `localkms` function: https://github.com/trustbloc/edge-core/issues/109.
+	kid, err := localkms.CreateKID(key.Value, kms.ED25519)
```

Comment:
I would pass the key type as an argument to Create()

Reply:
you can also sign using NIST P curves

Reply:
@Baha-sk 

`Create()` needs to satisfy an interface in `httpsignatures-go`